### ISSUE TITLE
Unable to create binary file command(make -C bin) in Cygwin using "make" command

### ISSUE THREAD

**Issue by Prabh95** [kaldi-asr/kaldi#1476]
I am using Cygwin on Windows 10. I as running the "make" command in cygwin for src folder in Kaldi. Before that while running make command, I had faced a few function declaration errors and I had corrected them. Now I am getting errors like these for the files I made declaration changes to: 

`make -C bin
make[2]: Entering directory '/cygdrive/c/Vocollect/kaldi-master/src/bin'
g++ -std=c++11 -I.. -I/cygdrive/c/Vocollect/kaldi-master/tools/openfst/include -Wno-sign-compare -Wall -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_CLAPACK -I../../tools/CLAPACK/ -msse -msse2 -g -O1 -DKALDI_PARANOID    -c -o align-equal.o align-equal.cc
g++   -g --enable-auto-import -L/usr/lib/lapack  align-equal.o ../decoder/kaldi-decoder.a ../lat/kaldi-lat.a ../lm/kaldi-lm.a ../fstext/kaldi-fstext.a ../hmm/kaldi-hmm.a ../transform/kaldi-transform.a ../gmm/kaldi-gmm.a ../tree/kaldi-tree.a ../util/kaldi-util.a ../thread/kaldi-thread.a ../matrix/kaldi-matrix.a ../base/kaldi-base.a   /cygdrive/c/Vocollect/kaldi-master/tools/openfst/lib/libfst.a -lcyglapack-0 -lcygblas-0 -lm -lpthread -ldl -o align-equal
align-equal.o: In function `kaldi::ParseOptions::ParseOptions(char const*)':
/cygdrive/c/Vocollect/kaldi-master/src/bin/../util/parse-options.h:43: undefined reference to `setlinebuf(__sFILE64*)'
/cygdrive/c/Vocollect/kaldi-master/src/bin/../util/parse-options.h:43:(.text$_ZN5kaldi12ParseOptionsC1EPKc[_ZN5kaldi12ParseOptionsC1EPKc]+0x1ee): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `setlinebuf(__sFILE64*)'
../util/kaldi-util.a(text-utils.o): In function `kaldi::IsToken(std::string const&)':

For example I had made change in the parse-options.h and added the declaration for setlinebuf function and the required library.

Can someone please help me create the binary file using make command in src?
`

**Comment by danpovey** [comment]
Please show the *original* error, before you attempted to correct anything.

**Comment by danpovey** [comment]
Hm.  From looking at various things online, it looks like the setlinebuf
function is not available in cygwin.
Anyway it's not strictly necessary, it will work without that.
The correct fix (which we should check in) is some kind of ifdef which will
prevent that function call from happening in cygwin.  But I can't figure
out the correct way (maybe #ifdef __CYGWIN__) ?
You could test that.

**Comment by Prabh95** [comment]
`make -C bin
make[2]: Entering directory '/cygdrive/c/Vocollect/kaldi-master/src/bin'
g++ -std=c++11 -I.. -I/cygdrive/c/Vocollect/kaldi-master/tools/openfst/include -Wno-sign-compare -Wall -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_CLAPACK -I../../tools/CLAPACK/ -msse -msse2 -g -O1 -DKALDI_PARANOID    -c -o align-equal.o align-equal.cc
g++   -g --enable-auto-import -L/usr/lib/lapack  align-equal.o ../decoder/kaldi-decoder.a ../lat/kaldi-lat.a ../lm/kaldi-lm.a ../fstext/kaldi-fstext.a ../hmm/kaldi-hmm.a ../transform/kaldi-transform.a ../gmm/kaldi-gmm.a ../tree/kaldi-tree.a ../util/kaldi-util.a ../thread/kaldi-thread.a ../matrix/kaldi-matrix.a ../base/kaldi-base.a   /cygdrive/c/Vocollect/kaldi-master/tools/openfst/lib/libfst.a -lcyglapack-0 -lcygblas-0 -lm -lpthread -ldl -o align-equal
align-equal.o: In function `kaldi::ParseOptions::ParseOptions(char const*)':
/cygdrive/c/Vocollect/kaldi-master/src/bin/../util/parse-options.h:43: undefined reference to `setlinebuf(__sFILE64*)'
/cygdrive/c/Vocollect/kaldi-master/src/bin/../util/**parse-options.h**:43:(.text$_ZN5kaldi12ParseOptionsC1EPKc[_ZN5kaldi12ParseOptionsC1EPKc]+0x1ee): relocation truncated to fit: R_X86_64_PC32 against undefined symbol **`setlinebuf(__sFILE64*)'**
../util/kaldi-util.a(text-utils.o): In function `kaldi::IsToken(std::string const&)':
/cygdrive/c/Vocollect/kaldi-master/src/util/**text-utils.cc:**110: undefined reference to **`isascii(int)'**
/cygdrive/c/Vocollect/kaldi-master/src/util/text-utils.cc:110:(.text+0x273): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `isascii(int)'
../util/kaldi-util.a(kaldi-io.o): In function `kaldi::PipeOutputImpl::Close()':
/cygdrive/c/Vocollect/kaldi-master/src/util/**kaldi-io.cc**:319: undefined reference to **`pclose(__sFILE64*)'**
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:319:(.text$_ZN5kaldi14PipeOutputImpl5CloseEv[_ZN5kaldi14PipeOutputImpl5CloseEv]+0x8f): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `pclose(__sFILE64*)'
../util/kaldi-util.a(kaldi-io.o): In function `kaldi::PipeInputImpl::Close()':
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:501: undefined reference to `pclose(__sFILE64*)'
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:501:(.text$_ZN5kaldi13PipeInputImpl5CloseEv[_ZN5kaldi13PipeInputImpl5CloseEv]+0x7a): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `pclose(__sFILE64*)'
../util/kaldi-util.a(kaldi-io.o): In function `kaldi::PipeOutputImpl::Open(std::string const&, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/util/**kaldi-io.cc:278**: undefined reference to **`popen(char const*,** char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:278:(.text$_ZN5kaldi14PipeOutputImpl4OpenERKSsb[_ZN5kaldi14PipeOutputImpl4OpenERKSsb]+0xa2): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `popen(char const*, char const*)'
../util/kaldi-util.a(kaldi-io.o): In function `kaldi::PipeInputImpl::Open(std::string const&, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:455: undefined reference to `popen(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:455:(.text$_ZN5kaldi13PipeInputImpl4OpenERKSsb[_ZN5kaldi13PipeInputImpl4OpenERKSsb]+0xae): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `popen(char const*, char const*)'
../util/kaldi-util.a(kaldi-io.o): In function `kaldi::PipeOutputImpl::Close()':
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:319: undefined reference to `pclose(__sFILE64*)'
/cygdrive/c/Vocollect/kaldi-master/src/util/kaldi-io.cc:319:(.text$_ZN5kaldi14PipeOutputImplD1Ev[_ZN5kaldi14PipeOutputImplD1Ev]+0x51): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `pclose(__sFILE64*)'
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::Matrix<float>::Init(int, int, kaldi::MatrixStrideType)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/**kaldi-matrix.cc**:661: undefined reference to **`posix_memalign**(void**, unsigned long, unsigned long)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:661:(.text$_ZN5kaldi6MatrixIfE4InitEiiNS_16MatrixStrideTypeE[_ZN5kaldi6MatrixIfE4InitEiiNS_16MatrixStrideTypeE]+0xb2): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::Matrix<double>::Init(int, int, kaldi::MatrixStrideType)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:661: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:661:(.text$_ZN5kaldi6MatrixIdE4InitEiiNS_16MatrixStrideTypeE[_ZN5kaldi6MatrixIdE4InitEiiNS_16MatrixStrideTypeE]+0xae): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::MatrixBase<float>::Invert(float*, float*, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:54: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:54:(.text$_ZN5kaldi10MatrixBaseIfE6InvertEPfS2_b[_ZN5kaldi10MatrixBaseIfE6InvertEPfS2_b]+0x111): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::MatrixBase<double>::Invert(double*, double*, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:54: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:54:(.text$_ZN5kaldi10MatrixBaseIdE6InvertEPdS2_b[_ZN5kaldi10MatrixBaseIdE6InvertEPdS2_b]+0x123): additional relocation overflows omitted from the output
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::Matrix<float>::Init(int, int, kaldi::MatrixStrideType)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:661: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-matrix.o):/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:569: more undefined references to `posix_memalign(void**, unsigned long, unsigned long)' follow
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::Matrix<double>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/**kaldi-matrix.cc**:1442: undefined reference to **strcasecmp**(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:1443: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:1446: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(kaldi-matrix.o): In function `kaldi::Matrix<float>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:1442: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:1443: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(kaldi-matrix.o):/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-matrix.cc:1446: more undefined references to `strcasecmp(char const*, char const*)' follow
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<float>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:181: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<double>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:181: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<float>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:181: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<double>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:181: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<double>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1189: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1190: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1193: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(kaldi-vector.o): In function `kaldi::Vector<float>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1189: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1190: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(kaldi-vector.o):/cygdrive/c/Vocollect/kaldi-master/src/matrix/kaldi-vector.cc:1193: more undefined references to `strcasecmp(char const*, char const*)' follow
../matrix/kaldi-matrix.a(packed-matrix.o): In function `kaldi::PackedMatrix<float>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:75: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(packed-matrix.o): In function `kaldi::PackedMatrix<double>::Init(int)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:75: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(packed-matrix.o): In function `kaldi::PackedMatrix<double>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:411: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:412: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:415: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(packed-matrix.o): In function `kaldi::PackedMatrix<float>::Read(std::istream&, bool, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:411: undefined reference to `strcasecmp(char const*, char const*)'
/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:412: undefined reference to `strcasecmp(char const*, char const*)'
../matrix/kaldi-matrix.a(packed-matrix.o):/cygdrive/c/Vocollect/kaldi-master/src/matrix/packed-matrix.cc:415: more undefined references to `strcasecmp(char const*, char const*)' follow
../matrix/kaldi-matrix.a(sp-matrix.o): In function `kaldi::SpMatrix<float>::Invert(float*, float*, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/sp-matrix.cc:253: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
../matrix/kaldi-matrix.a(sp-matrix.o): In function `kaldi::SpMatrix<double>::Invert(double*, double*, bool)':
/cygdrive/c/Vocollect/kaldi-master/src/matrix/sp-matrix.cc:253: undefined reference to `posix_memalign(void**, unsigned long, unsigned long)'
**collect2: error: ld returned 1 exit status
make[2]: *** [<builtin>: align-equal] Error 1
make[2]: Leaving directory '/cygdrive/c/Vocollect/kaldi-master/src/bin'
make[1]: *** [Makefile:142: bin] Error 2
make[1]: Leaving directory '/cygdrive/c/Vocollect/kaldi-master/src'
make: *** [Makefile:38: all] Error 2**
`

**Comment by Prabh95** [comment]
See this is the whole error that I am getting while trying to create a binary file. In short it contains all the functions like isascii,strcasecmp,posix_memalign, popen, pclose etc. to which I had to add their function definition in the required files in order to remove the earlier errors of function not declared.

**Comment by danpovey** [comment]
Those are system functions, you should never have to add declarations for
them, and definitely not definitions.
It's almost certainly a problem with the way the compiler was invoked, or a
problem with the compiler installation itself.  I'm afraid there is not
much I can do to help in the short term.  I don't have access to any
Windows machines.  Yenda might look at it, but probably it compiles fine
for him so I don't know how we'd fix it if we can't reproduce it.
You'll have to look for similar errors online.
Unfortunately cygwin is not the most stable platform, it doesn't get as
much use as Linux so it's never as thoroughly debugged.   (in general, not
just for Kaldi).

**Comment by Prabh95** [comment]
Firstly thankyou so much for your active responses on the issue. In the message above I was talking about function declarations(Sorry, I accidently wrote function definition). 

Okay I removed all the declarations which I had made. I run the commands for configure and make depend which run successfully in SRC.

But when I try to run Make, I face errors like these while compiling on cygwin

`In file included from ../base/kaldi-common.h:34:0,
                 from ../matrix/matrix-common.h:26,
                 from ../matrix/kaldi-matrix.h:26,
                 from kaldi-matrix.cc:23:
kaldi-matrix.cc: In member function ‘void kaldi::MatrixBase<Real>::Invert(Real*, Real*, bool)’:
../base/kaldi-utils.h:58:43: error: there are no arguments to ‘posix_memalign’ that depend on a template parameter, so a declaration of ‘posix_memalign’ must be available [-fpermissive]
      (!posix_memalign(pp_orig, align, size) ? *(pp_orig) : NULL)
                                           ^
kaldi-matrix.cc:54:11: note: in expansion of macro ‘KALDI_MEMALIGN’
           KALDI_MEMALIGN(16, sizeof(Real)*l_work, &temp))) == NULL)
           ^
../base/kaldi-utils.h:58:43: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
      (!posix_memalign(pp_orig, align, size) ? *(pp_orig) : NULL)


I get similar errors for functions isascii,strcasecmp, popen, pclose, setlinebuf.
Please help! How do I resolve this?
`

**Comment by danpovey** [comment]
Hm, it used to work, cygwin must have changed.  You could try typing "man
posix_memalign" and similarly for the other functions, and see where they
are declared; and add suitable #include statements.

**Comment by Prabh95** [comment]
I tried including the header files accordingly, didnt help sir. Anything else that can be done? or should I use Ubuntu then?

**Comment by danpovey** [comment]
Yes, use Ubuntu.  Next week we may have time to check that Kaldi is
compiling on cygwin.

**Comment by Prabh95** [comment]
Okay I was using Kaldi in linux environment. I ran the recipe for yesno(which contains 8000hz .wav files) and then I modified it by adding a sample .wav file recorded at 16,000hz to the Yesno dataset. The file wasn't read obviously but it didnt show any error for the 16,000hz file. And it read the files until that 16khz file was to be read and the rest of the files after that weren't trained or tested. I wanted to ask why no error was displayed?
